### PR TITLE
Ihamer/7468 inject noc delays

### PR DESCRIPTION
**Pull request by ihamer-tt**
For issue [#7468](https://github.com/tenstorrent/tt-metal/issues/7468).

### CODE REVIEW

**Comment by tt-rkim** [comment]
Let us know when this is ready.

Looks like it contains changes needed from Tracy cc: @mo-tenstorrent

File: tt_metal/third_party/tracy
Comment (mo-tenstorrent) [inline tt_metal/third_party/tracy:1]: This is a reversion to the older version of tracy. Could be an artifact of force pushing before `submodule update`

**Comment by ihamer-tt** [comment]
Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/9270563547

**Review by TT-billteng — APPROVED** [review]

File: docs/source/tt-metalium/tools/watcher.rst [hunk 1]
```diff
@@ -201,3 +201,24 @@ watcher log:
          0x00000020,0x0000001f,0x0000001e,0x0000001d,0x0000001c,0x0000001b,0x0000001a,0x00000019,
          0x00000018,0x00000017,0x00000016,0x00000015,0x00000014,0x00000013,0x00000012,0x00000011,
          0x00000010,0x0000000f,0x0000000e,0x0000000d,0x0000000c,0x0000000b,0x0000000a]
+
+
+Debug Delays
+------------
+Watcher can insert NOC transactions delays for debugging purposes. The delayes can be specified by
```
Comment (tt-dma) [inline docs/source/tt-metalium/tools/watcher.rst:208]:
```suggestion
Watcher can insert NOC transaction delays for debugging purposes. These delays can be specified by
```

File: docs/source/tt-metalium/tools/watcher.rst [hunk 2]
```diff
@@ -201,3 +201,24 @@ watcher log:
...
+is applied to all RISCs.
+Note that `TT_METAL_WATCHER` must be set and `TT_METAL_WATCHER_DISABLE_NOC_SANITIZE` must not be
+set for the delays to be applied.
+
+For example, the following environment variables will add a delay of 10 iterations for both READ
+and WRITE transactions on BRISC core at location 0,0.
+
+.. code-block::
+
+    TT_METAL_WATCHER=1 TT_METAL_WATCHER_DEBUG_DELAY=10 TT_METAL_READ_DEBUG_DELAY_CORES=0,0 TT_METAL_WRITE_DEBUG_DELAY_CORES=0,0 TT_METAL_READ_DEBUG_DELAY_RISCVS=BR TT_METAL_WRITE_DEBUG_DELAY_RISCVS=BR
```
Comment (tt-rkim) [inline docs/source/tt-metalium/tools/watcher.rst:224]: Did you check on a local docs build that these environment display on new lines? I'm unsure if that's your intention.

I would also suggest writing an example command with a unit tests executable to show how to use it. If it's one thing I've learned, it's never assume your readers know what you're talking about, including as something elementary as Unix environment variables.

File: tt_metal/impl/debug/watcher_server.cpp [hunk 3]
```diff
@@ -899,17 +888,108 @@ void watcher_init(Device *device) {
...
+            }
+
+            for (CoreType core_type : {CoreType::WORKER, CoreType::ETH}) {
+                vector<CoreCoord> delayed_cores = tt::llrt::OptionsG.get_feature_cores(delay_feature)[core_type];
+                for (tt_xy_pair logical_core : delayed_cores) {
+                    CoreCoord phys_core;
+                    bool valid_logical_core = true;
+                    try {
+                        phys_core = device->physical_core_from_logical_core(logical_core, core_type);
+                    } catch (std::runtime_error &error) {
```
Comment (tt-dma) [inline tt_metal/impl/debug/watcher_server.cpp:925]: Should we print a message saying that the given core doesn't map to anything and will be ignored? We do that for dprint

**Review by tt-rkim — APPROVED** [review]

File: tt_metal/hw/inc/debug/sanitize_noc.h [hunk 4]
```diff
@@ -209,61 +186,102 @@ void debug_sanitize_noc_and_worker_addr(
...
+    bool delay = false;
+    switch (transaction_type) {
+        case TransactionRead: delay = (v[0].read_delay_riscv_mask & (1 << debug_get_which_riscv())) != 0; break;
+        case TransactionWrite: delay = (v[0].write_delay_riscv_mask & (1 << debug_get_which_riscv())) != 0; break;
+        case TransactionAtomic: delay = (v[0].atomic_delay_riscv_mask & (1 << debug_get_which_riscv())) != 0; break;
+        default: break;
+    }
+    if (delay) {
+        // WATCHER_DEBUG_DELAY is a compile time constant passed with -D
+        for (volatile uint32_t i = 0; i < WATCHER_DEBUG_DELAY; i++);  // Spin loop
```
Comment (tt-dma) [inline tt_metal/hw/inc/debug/sanitize_noc.h:265]: Should we consider using `riscv_wait()` function here to use the wallclock?

**Comment by tt-dma** [comment]
Can we just update the docs to say `cycles` instead of `iterations` now that we're using `riscv_wait()`? Aside from that looks good to me

**Review by tt-dma — APPROVED** [review]

**Pull request merged**